File: backend/src/main/scala/services/battle/services/world/BattleMotionRules.scala
package services.battle.services.world

import services.battle.services.*

import services.battle.objects.*
import services.battle.services.world.BattleArenaCollision.*
import services.battle.services.world.BattleGeometry.*

private[services] object BattleMotionRules {
  final case class SteppedMotionResult(
    destination: BattleVector2,
    blocked: Boolean,
    hitBlocker: Boolean
  )

  private final case class SteppedMotionScan(
    lastValid: BattleVector2,
    hitBlocker: Boolean
  )

  /** 中文名：规范化移动（normalizeMovement）。游戏职责：在后端世界域中管理地图、碰撞、几何、移动和出生点，约束实体在战场中的空间行为。 */
  def normalizeMovement(next: BattleVector2): BattleVector2 = {
    val length = math.hypot(next.x, next.y)
    if length <= 0.0001 then BattleArenaCatalog.ZeroVector
    else BattleVector2(next.x / length, next.y / length)
  }

  /** 中文名：查找运动destination（findMotionDestination）。游戏职责：在后端世界域中管理地图、碰撞、几何、移动和出生点，约束实体在战场中的空间行为。 */
  def findMotionDestination(
    position: BattleVector2,
    direction: BattleVector2,
    distance: Double,
    radius: Double
  ): SteppedMotionResult = {
    val normalized = normalizeMovement(direction)
    val clampedDistance = math.max(0.0, distance)
    val fullMotion = resolveSteppedMotion(position, normalized, clampedDistance, radius)

    if !fullMotion.hitBlocker then fullMotion
    else {
      val xDistance = math.abs(normalized.x * clampedDistance)
      val yDistance = math.abs(normalized.y * clampedDistance)
      val xMotion =
        if xDistance > 0.0 then resolveSteppedMotion(position, BattleVector2(math.signum(normalized.x), 0.0), xDistance, radius)
        else fullMotion
      val yMotion =
        if yDistance > 0.0 then resolveSteppedMotion(position, BattleVector2(0.0, math.signum(normalized.y)), yDistance, radius)
        else fullMotion

      Vector(fullMotion, xMotion, yMotion).maxBy(motion => distanceBetween(position, motion.destination))
    }
  }

  private def resolveSteppedMotion(
    position: BattleVector2,
    direction: BattleVector2,
    distance: Double,
    radius: Double
  ): SteppedMotionResult = {
    val clampedDistance = math.max(0.0, distance)
    val steps = math.ceil(clampedDistance / BattleArenaCatalog.MotionStepSize).toInt
    val scan = (1 to steps).foldLeft(SteppedMotionScan(position, hitBlocker = false)) { (current, step) =>
      if current.hitBlocker then current
      else {
        val travel = math.min(clampedDistance, step.toDouble * BattleArenaCatalog.MotionStepSize)
        val candidate = add(position, scale(direction, travel))
        if canPlayerOccupy(candidate, radius) then current.copy(lastValid = candidate)
        else current.copy(hitBlocker = true)
      }
    }

    SteppedMotionResult(
      destination = scan.lastValid,
      blocked = scan.lastValid == position,
      hitBlocker = scan.hitBlocker
    )
  }
}

File: frontend/src/runtime/battle/microservices/world/functions/BattleMotionRules.ts
import type { BattleVector2 as Vec2 } from "../../../../../objects/battle/objects/core/BattleCoreScalars";
import {
  canPlayerOccupy,
  collidesWithArenaObstacles,
  intersectsObstacle,
  isInWorld,
  type BattleArenaObstacleBounds,
  type BattleArenaObstacleShape
} from "./BattleArenaCollision";

export type MotionObstacleBounds = BattleArenaObstacleBounds;
export type MotionObstacleShape = BattleArenaObstacleShape;

export interface MotionDestinationInput {
  position: Vec2;
  direction: Vec2;
  distance: number;
  radius: number;
  worldSize: Vec2;
  obstacleBounds: readonly MotionObstacleBounds[];
}

export interface MotionTargetInput {
  target: Vec2;
  radius: number;
  worldSize: Vec2;
  obstacleBounds: readonly MotionObstacleBounds[];
}

export interface MotionControllerTargetResult {
  destination: Vec2;
  blocked: boolean;
}

interface SteppedMotionResult extends MotionControllerTargetResult {
  hitBlocker: boolean;
}

export function normalizeMovement(vector: Vec2): Vec2 {
  const length = Math.hypot(vector.x, vector.y);
  if (length <= 0.0001) {
    return { x: 0, y: 0 };
  }

  return {
    x: vector.x / length,
    y: vector.y / length
  };
}

export function normalizeVector(vector: Vec2): Vec2 {
  return normalizeMovement(vector);
}

export function findMotionDestination(input: MotionDestinationInput): MotionControllerTargetResult {
  const normalized = normalizeMovement(input.direction);
  const distance = Math.max(0, input.distance);
  const fullMotion = tryResolveSteppedMotion(
    input.position,
    normalized,
    distance,
    input.radius,
    input.worldSize,
    input.obstacleBounds
  );

  if (!fullMotion.hitBlocker) {
    return toMotionResult(fullMotion);
  }

  const xDistance = Math.abs(normalized.x * distance);
  const yDistance = Math.abs(normalized.y * distance);
  const xMotion =
    xDistance > 0
      ? tryResolveSteppedMotion(
          input.position,
          { x: Math.sign(normalized.x), y: 0 },
          xDistance,
          input.radius,
          input.worldSize,
          input.obstacleBounds
        )
      : fullMotion;
  const yMotion =
    yDistance > 0
      ? tryResolveSteppedMotion(
          input.position,
          { x: 0, y: Math.sign(normalized.y) },
          yDistance,
          input.radius,
          input.worldSize,
          input.obstacleBounds
        )
      : fullMotion;

  return toMotionResult(resolveBestMotion(input.position, fullMotion, xMotion, yMotion));
}

function tryResolveSteppedMotion(
  position: Vec2,
  direction: Vec2,
  distance: number,
  radius: number,
  worldSize: Vec2,
  obstacleBounds: readonly MotionObstacleBounds[]
): SteppedMotionResult {
  let lastValid = { x: position.x, y: position.y };
  const clampedDistance = Math.max(0, distance);
  const steps = Math.ceil(clampedDistance / 16);
  let hitBlocker = false;

  for (let step = 1; step <= steps; step += 1) {
    const travel = Math.min(clampedDistance, step * 16);
    const candidate = {
      x: position.x + direction.x * travel,
      y: position.y + direction.y * travel
    };

    if (!canPlayerOccupy(candidate, radius, worldSize, obstacleBounds)) {
      hitBlocker = true;
      break;
    }

    lastValid = candidate;
  }

  return {
    destination: lastValid,
    blocked: lastValid.x === position.x && lastValid.y === position.y,
    hitBlocker
  };
}

function toMotionResult(result: SteppedMotionResult): MotionControllerTargetResult {
  return {
    destination: result.destination,
    blocked: result.blocked
  };
}

function motionProgress(destination: Vec2, origin: Vec2): number {
  return Math.hypot(destination.x - origin.x, destination.y - origin.y);
}

function resolveBestMotion(origin: Vec2, ...motions: readonly SteppedMotionResult[]): SteppedMotionResult {
  return motions.reduce((bestMotion, motion) =>
    motionProgress(motion.destination, origin) > motionProgress(bestMotion.destination, origin) ? motion : bestMotion
  );
}

export function isMotionTargetPointValid(input: MotionTargetInput): boolean {
  return canPlayerOccupy(input.target, input.radius, input.worldSize, input.obstacleBounds);
}

export function canOccupy(
  position: Vec2,
  radius: number,
  worldSize: Vec2,
  obstacleBounds: readonly MotionObstacleBounds[]
): boolean {
  return canPlayerOccupy(position, radius, worldSize, obstacleBounds);
}

export function collidesWithObstacles(
  position: Vec2,
  radius: number,
  obstacleBounds: readonly MotionObstacleBounds[]
): boolean {
  return collidesWithArenaObstacles(position, radius, obstacleBounds);
}

export { intersectsObstacle };
export const isInsideWorld = isInWorld;
